File: examples/ios/Examples/ManageApiKeys.m
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>
#import "MyRealmApp.h"

@interface ManageApiKeys : XCTestCase

@end

@implementation ManageApiKeys

- (void)setUp {
    XCTestExpectation *expectation = [self expectationWithDescription:@"registers and logs in"];

    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID]; // Replace YOUR_REALM_APP_ID with your Realm app ID
    RLMEmailPasswordAuth *client = [app emailPasswordAuth];
    NSString *email = @"manage-api-keys-objc@example.com";
    NSString *password = @"123456";
    // User will be deleted by TestSetup after entire suite
    [client registerUserWithEmail:email password:password completion:^(NSError *error) {
        // Ignore error. May have registered on previous test.
        [app loginWithCredential:[RLMCredentials credentialsWithEmail:email password:password] completion:^(RLMUser *user, NSError *error) {
                [expectation fulfill];
            }];
    }];
    // :code-block-end:
    [self waitForExpectations:@[expectation] timeout:10.0];
}

- (void)tearDown {
    XCTestExpectation *expectation = [self expectationWithDescription:@"Log out"];
    [[[MyRealmApp app] currentUser] logOutWithCompletion:^(NSError *error) {
        [expectation fulfill];
    }];
    [self waitForExpectations:@[expectation] timeout:10.0];
}

- (void)testCreateApiKey {
    XCTestExpectation *expectation = [self expectationWithDescription:@"it completes"];
    // :code-block-start: create-api-key
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    // ... log in ...
    RLMUser *user = [app currentUser];
    RLMAPIKeyAuth *client = [user apiKeysAuth];

    // Create the API key
    [client createAPIKeyWithName:@"someKeyName" completion:^(RLMUserAPIKey *apiKey, NSError *error) {
        if (error != nil) {
            // ... handle Error ...
        } else {
            // ... use apiKey ...
            // :hide-start:
            [client deleteAPIKey:[apiKey objectId] completion:^(NSError *error) {
                [expectation fulfill];
            }];
            // :hide-end:
        }
    }];
    // :code-block-end:
    [self waitForExpectations:@[expectation] timeout:10.0];
}

- (void)testLookUpApiKey {
    XCTestExpectation *fetchOneExpectation = [self expectationWithDescription:@"fetch one completes"];
    XCTestExpectation *fetchAllExpectation = [self expectationWithDescription:@"fetch all completes"];
    // :code-block-start: look-up-api-key
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    // ... log in ...
    RLMUser *user = [app currentUser];
    RLMAPIKeyAuth *client = [user apiKeysAuth];

    // Fetch API key by a specific ObjectId
    NSError *error = nil;
    RLMObjectId *objectId = [[RLMObjectId alloc] initWithString:@"someObjectId" error:&error];
    [client fetchAPIKey:objectId completion:^(RLMUserAPIKey *apiKey, NSError *error) {
       if (error != nil) {
          // ... handle error ...
       } else {
          // ... use apiKey ...
       }
       // :hide-start:
       [fetchOneExpectation fulfill];
       // :hide-end:
    }];

    // Fetch all API keys
    [client fetchAPIKeysWithCompletion:^(NSArray<RLMUserAPIKey *> *keys, NSError *error) {
       if (error != nil) {
          // ... handle error ...
       } else {
          for (RLMUserAPIKey *key in keys) {
                // ... use key ...
          }
       }
       // :hide-start:
       [fetchAllExpectation fulfill];
       // :hide-end:
    }];
    // :code-block-end:
    [self waitForExpectations:@[fetchOneExpectation, fetchAllExpectation] timeout:10.0];
}

- (void)testEnableDisableApiKey {
    XCTestExpectation *enableExpectation = [self expectationWithDescription:@"enable completes"];
    XCTestExpectation *disableExpectation = [self expectationWithDescription:@"disable completes"];
    // :code-block-start: enable-disable-api-key
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    // ... log in ...
    RLMUser *user = [app currentUser];
    RLMAPIKeyAuth *client = [user apiKeysAuth];
    
    // Enable the API key with specific objectId
    RLMObjectId *objectId = [[RLMObjectId alloc] initWithString:@"00112233445566778899aabb" error:nil];
    [client enableAPIKey:objectId completion:^(NSError *error) {
       // Handle error if any. Otherwise, enable was successful.
       // :hide-start:
       [enableExpectation fulfill];
       // :hide-end:
    }];
    
    RLMUserAPIKey *apiKey;

    // ... Get an API key ...
    // :hide-start:
    apiKey = [[RLMUserAPIKey alloc] init];
    // :hide-end:

    // Disable the API key
    [client disableAPIKey:[apiKey objectId] completion:^(NSError *error) {
       // Handle error if any. Otherwise, disable was successful.
       // :hide-start:
       [disableExpectation fulfill];
       // :hide-end:
    }];
    // :code-block-end:
    [self waitForExpectations:@[enableExpectation, disableExpectation] timeout:10.0];
}

- (void)testDeleteApiKey {
    XCTestExpectation *expectation = [self expectationWithDescription:@"it completes"];
    // :code-block-start: delete-api-key
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    // ... log in ...
    RLMUser *user = [app currentUser];
    RLMAPIKeyAuth *client = [user apiKeysAuth];
    
    RLMUserAPIKey *apiKey;

    // ... Get an API key ...
    // :hide-start:
    apiKey = [[RLMUserAPIKey alloc] init];
    // :hide-end:

    [client deleteAPIKey:[apiKey objectId] completion:^(NSError *error) {
       // Handle error if any. Otherwise, delete was successful.
       // :hide-start:
       [expectation fulfill];
       // :hide-end:
    }];
    // :code-block-end:
    [self waitForExpectations:@[expectation] timeout:10.0];
}

@end

File: examples/ios/Examples/ManageApiKeys.swift
import XCTest
import RealmSwift

class ManageApiKeys: XCTestCase {
    override func setUp() {
        let expectation = XCTestExpectation(description: "logs in")
        let email = "api-key-user@example.com"
        let password = "123456"

        // User will be deleted by TestSetup after entire suite 
        app.emailPasswordAuth.registerUser(email: email, password: password) { (error) in
            // Ignore "name already in use" error as it may have been created by a previous test in this suite
            guard error == nil || error!.localizedDescription == "name already in use" else {
                fatalError("User registration failed: \(error!.localizedDescription)")
            }
            app.login(credentials: Credentials.emailPassword(email: email, password: password)) { (result) in
                switch result {
                case .failure(let error):
                    fatalError("Login failed: \(error.localizedDescription)")
                case .success(_):
                    expectation.fulfill()
                }
            }
        }

        wait(for: [expectation], timeout: 10)
    }

    override func tearDown() {
        let expectation = XCTestExpectation(description: "logs out")
        app.currentUser!.logOut { (error) in
            guard error == nil else {
                fatalError("Failed to log out: \(error!.localizedDescription)")
            }
            expectation.fulfill()
        }
        wait(for: [expectation], timeout: 10)
    }

    func testCreateApiKey() {
        let expectation = XCTestExpectation(description: "it completes")
        // :code-block-start: create-api-key
        let app = App(id: YOUR_REALM_APP_ID)

        // ... log in ...

        // User must not be an anonymous user.
        let user = app.currentUser!
        let client = user.apiKeysAuth

        client.createAPIKey(named: "someKeyName") { (apiKey, error) in
            guard error == nil else {
                print("Failed to create key: \(error!)")
                return
            }
            // Use apiKey
            // :hide-start:
            client.deleteAPIKey(apiKey!.objectId) { (error) in
                expectation.fulfill()
            }
            // :hide-end:
        }

        // :code-block-end:
        wait(for: [expectation], timeout: 10)
    }

    func testLookUpApiKey() {
        let singleFetchExpectation = XCTestExpectation(description: "single fetch completes")
        let fetchAllExpectation = XCTestExpectation(description: "fetch all completes")
        // :code-block-start: look-up-api-key
        let app = App(id: YOUR_REALM_APP_ID)

        // ... log in ...

        let user = app.currentUser!
        let client = user.apiKeysAuth

        // Fetch a specific API key by ObjectId
        client.fetchAPIKey(ObjectId("00112233445566778899aabb")) { (maybeApiKey, error) in
            // ...
            // :hide-start:
            singleFetchExpectation.fulfill()
            // :hide-end:
        }

        // Fetch all API keys
        client.fetchAPIKeys { (keys, error) in
            guard error == nil else {
                fatalError("Failed to fetch keys: \(error!)")
            }
            for key in keys! {
                // use key
                print(key.name)
            }
            // :hide-start:
            fetchAllExpectation.fulfill()
            // :hide-end:
        }
        // :code-block-end:
        wait(for: [singleFetchExpectation, fetchAllExpectation], timeout: 10)
    }

    func testEnableDisableApiKey() {
        let enableExpectation = XCTestExpectation(description: "enable completes")
        let disableExpectation = XCTestExpectation(description: "disable completes")

        // :code-block-start: enable-disable-api-key
        let app = App(id: YOUR_REALM_APP_ID)

        // ... log in ...

        let user = app.currentUser!

        let client = user.apiKeysAuth

        // Enable the API key
        client.enableAPIKey(ObjectId("00112233445566778899aabb")) { (error) in
            // ...
            // :hide-start:
            enableExpectation.fulfill()
            // :hide-end:
        }

        let apiKey: UserAPIKey?
        
        // ... Obtain a user API key ...
        // :hide-start:
        apiKey = UserAPIKey()
        // :hide-end:
        
        // Disable the API key
        client.disableAPIKey(apiKey!.objectId) { (error) in
            // ...
            // :hide-start:
            disableExpectation.fulfill()
            // :hide-end:
        }

        // :code-block-end:
        wait(for: [enableExpectation, disableExpectation], timeout: 10)
    }

    func testDeleteApiKey() {
        let expectation = XCTestExpectation(description: "it completes")
        // :code-block-start: delete-api-key
        let app = App(id: YOUR_REALM_APP_ID)

        // ... log in ...

        let user = app.currentUser!
        let client = user.apiKeysAuth
        
        let apiKey: UserAPIKey?
        
        // ... Obtain a user API key ...
        // :hide-start:
        apiKey = UserAPIKey()
        // :hide-end:
        
        client.deleteAPIKey(apiKey!.objectId) { (error) in
            guard error == nil else {
                print("Failed to delete key: \(error!)")
                // :hide-start:
                XCTAssertEqual(error!.localizedDescription, "API key not found")
                expectation.fulfill()
                // :hide-end:
                return
            }
            // Key deleted
        }
        // :code-block-end:

        wait(for: [expectation], timeout: 10)
    }
}
